lib/seahorn/BvOpSem2.cc

```diff
@@ -2386,11 +2386,37 @@ Bv2OpSemContext::Bv2OpSemContext(SymStore &values, ExprVector &side,
   setPathCond(o.getPathCond());
 }
 
+Expr Bv2OpSemContext::convertConstArrayToLambda(Expr e) {
+  // go into structs until a non struct is found; convert to lambda if needed.
+  if (strct::isStructVal(e)) {
+    llvm::SmallVector<Expr, 8> kids;
+    for (unsigned i = 0, sz = e->arity(); i < sz; ++i) {
+
+      kids.push_back(convertConstArrayToLambda(e->arg(i)));
+    }
+    return strct::mk(kids);
+  } else if (isOpX<CONST_ARRAY>(e)) {
+    Expr dom = e->arg(0); // get domain
+    Expr val = e->arg(1); // get value
+    Expr addr = bind::mkConst(mkTerm<std::string>("addr", efac()), dom);
+    Expr decl = bind::fname(addr);
+    Expr res = mk<LAMBDA>(decl, val);
+    return res;
+  } else {
+    return e;
+  }
+}
+
 Expr Bv2OpSemContext::simplify(Expr u) {
   ScopedStats _st_("opsem.simplify");
 
-  Expr _u;
-  _u = m_z3_simplifier->simplify(u);
+  Expr __u, _u;
+  __u = m_z3_simplifier->simplify(u);
+  if (UseLambdas) {
+    _u = convertConstArrayToLambda(__u);
```

Comment:
will be more readable as a ternary ite operation
```
_u = UseLambdas ? ... 
```

Reply:
good point! done.